Local Socket API support

chasenicholl opened the pull request:
### [v4.0.0]

Added Local UDP API support! Now you can choose to listen to your Weather Stations observations directly over your local network. No Station ID or API Token needed. Observations are broadcasted every 60 seconds. This leverages the obs_st message. See [documentation](https://weatherflow.github.io/Tempest/api/udp/v171/) for more information.

chasenicholl committed:
checkpoint.

chasenicholl committed:
local api support

chasenicholl committed:
update changelog

chasenicholl committed:
updated wait function

chasenicholl commented:
@DMBlakeley take a look when you get a chance - local udp API implementation!

DMBlakeley commented:
@chasenicholl will take a look. Nice work!

Have the 4.0.0 version running on my MacBook and 3.0.3 version running on my Mac mini so I can easily compare between the two versions.

chasenicholl commented:
Thanks @DMBlakeley!

DMBlakeley committed:
Local UDP API support

* Added Local UDP API support! Now you can choose to listen to your Weather Stations observations directly over your local network. No Station ID or API Token needed. Observations are broadcasted every 60 seconds. This leverages the `obs_st` message. See [documentation](https://weatherflow.github.io/Tempest/api/udp/v171/) for more information.

DMBlakeley commented:
Hi @chasenicholl,

I have dug into v4.0.0 comparing it to v3.0.3 as well as the Tempest App directly. I have a number of recommended changes to a number of files:

- `config.schema.json`: remove `!model.local_api && ` from sensor properties. Still need `functionBody` for local API.
- `README.md`: update Local API example to include `functionBody`
- `platformAccessory.ts`: picked up an error in the `getCurrentWindSpeed()` function for windspeed that existed in v3.0.3.
- `tempest.ts`: added equations for `heatIndex`, `feelsLike` and `windChill`. Corrected units.

For the `tempest.ts` changes I was reminded how HomeKit handles metric vs standard units. For supported units such as `temperature`, `humidity`, and `wind speed` calculations are done with metric units. How they are displayed is a function of the localization of your iOS or Mac device. The units for additional parameters that were added as dummy occupancy sensor are handled by the plugin settings.

All the updated files can be found in my directory ([https://github.com/DMBlakeley/homebridge-weatherflow-tempest](https://github.com/DMBlakeley/homebridge-weatherflow-tempest)). Now need to figure out how to push to you.

DMBlakeley commented on config.schema.json:
```diff
@@ -1,182 +1,214 @@
+                          "title": "Fan Properties",
+                          "type": "object",
+                          "condition": {
+                            "functionBody": "if (!model.local_api && model.sensors[arrayIndices] && model.sensors[arrayIndices].sensor_type && model.sensors[arrayIndices].sensor_type === 'Fan') { return true; } else { return false; };"
```
Delete `!model.local_api && ` as `functionBody` applies to both Local and Cloud APIs.

DMBlakeley commented on config.schema.json:
```diff
@@ -1,182 +1,214 @@
+                          "title": "Light Properties",
+                          "type": "object",
+                          "condition": {
+                            "functionBody": "if (!model.local_api && model.sensors[arrayIndices] && model.sensors[arrayIndices].sensor_type && model.sensors[arrayIndices].sensor_type === 'Light Sensor') { return true; } else { return false; };"
```
Delete `!model.local_api && ` as `functionBody` applies to both Local and Cloud APIs.

DMBlakeley commented on config.schema.json:
```diff
@@ -1,182 +1,214 @@
+                          "title": "Humidity Properties",
+                          "type": "object",
+                          "condition": {
+                            "functionBody": "if (!model.local_api && model.sensors[arrayIndices] && model.sensors[arrayIndices].sensor_type && model.sensors[arrayIndices].sensor_type === 'Humidity Sensor') { return true; } else { return false; };"
```
Delete `!model.local_api && ` as `functionBody` applies to both Local and Cloud APIs.

DMBlakeley commented on config.schema.json:
```diff
@@ -1,182 +1,214 @@
+                          "title": "Temperature Properties",
+                          "type": "object",
+                          "condition": {
+                            "functionBody": "if (!model.local_api && model.sensors[arrayIndices] && model.sensors[arrayIndices].sensor_type && model.sensors[arrayIndices].sensor_type === 'Temperature Sensor') { return true; } else { return false; };"
```
Delete `!model.local_api && ` as `functionBody` applies to both Local and Cloud APIs.

DMBlakeley commented on config.schema.json:
```diff
@@ -1,182 +1,214 @@
+                          "title": "Motion Properties",
+                          "type": "object",
+                          "condition": {
+                            "functionBody": "if (model.sensors[arrayIndices] && model.sensors[arrayIndices].sensor_type && model.sensors[arrayIndices].sensor_type === 'Motion Sensor') { return true; } else { return false; };"
```
Delete `!model.local_api && ` as `functionBody` applies to both Local and Cloud APIs.

DMBlakeley commented on config.schema.json:
```diff
@@ -1,182 +1,214 @@
+                                      "wind_gust"
+                                  ],
+                                  "condition": {
+                                      "functionBody": "if (!model.local_api && model.sensors[arrayIndices] && model.sensors[arrayIndices].sensor_type && model.sensors[arrayIndices].sensor_type === 'Temperature Sensor') { return true; } else { return false; };"
```
Delete `!model.local_api && ` as `functionBody` applies to both Local and Cloud APIs.

DMBlakeley commented on config.schema.json:
```diff
@@ -1,182 +1,214 @@
+                                      "uv"
+                                  ],
+                                  "condition": {
+                                      "functionBody": "if (!model.local_api && model.sensors[arrayIndices] && model.sensors[arrayIndices].sensor_type && model.sensors[arrayIndices].sensor_type === 'Temperature Sensor') { return true; } else { return false; };"
```
Delete `!model.local_api && ` as `functionBody` applies to both Local and Cloud APIs.
Change `Temperature Sensor` to `Occupancy Sensor`

DMBlakeley commented on README.md:
```diff
@@ -69,11 +72,90 @@ sensor_type `{2}` | value_key | metric units | std units | additional_properties
+  "name": "WeatherFlow Tempest Platform",
+  "local_api": true,
+  "units": "Standard",
+  "sensors": [
```
Replace lines 82 to 148 with to reflect updates to `config.schema.json`:

`
"sensors": [
      {
          "name": "Temperature",
          "sensor_type": "Temperature Sensor",
          "temperature_properties": {
              "value_key": "air_temperature"
          }
      },
      {
          "name": "Feels Like",
          "sensor_type": "Temperature Sensor",
          "temperature_properties": {
              "value_key": "feels_like"
          }
      },
      {
          "name": "Dew Point",
          "sensor_type": "Temperature Sensor",
          "temperature_properties": {
              "value_key": "dew_point"
          }
      },
      {
          "name": "Relative Humidity",
          "sensor_type": "Humidity Sensor",
          "humidity_properties": {
              "value_key": "relative_humidity"
          }
      },
      {
          "name": "Light Level",
          "sensor_type": "Light Sensor",
          "light_properties": {
              "value_key": "brightness"
          }
      },
      {
          "name": "Wind Speed",
          "sensor_type": "Fan",
          "fan_properties": {
              "value_key": "wind_avg"
          }
      },
      {
          "name": "Wind Gust",
          "sensor_type": "Motion Sensor",
          "motion_properties": {
              "value_key": "wind_gust",
              "trigger_value": 10
          }
      },
      {
          "name": "Barometer",
          "sensor_type": "Occupancy Sensor",
          "occupancy_properties": {
              "value_key": "barometric_pressure",
              "trigger_value": 30
          }
      },
      {
          "name": "Solar Radiation",
          "sensor_type": "Occupancy Sensor",
          "occupancy_properties": {
              "value_key": "solar_radiation",
              "trigger_value": 1000
          }
      },
      {
          "name": "UV",
          "sensor_type": "Occupancy Sensor",
          "occupancy_properties": {
              "value_key": "uv",
              "trigger_value": 3
          }
      },
      {
          "name": "Precipitation Rate",
          "sensor_type": "Occupancy Sensor",
          "occupancy_properties": {
              "value_key": "precip",
              "trigger_value": 0.25
          }
      },
      {
          "name": "Precipitation Today",
          "sensor_type": "Occupancy Sensor",
          "occupancy_properties": {
              "value_key": "precip_accum_local_day",
              "trigger_value": 1
          }
      },
      {
          "name": "Wind Direction",
          "sensor_type": "Occupancy Sensor",
          "occupancy_properties": {
              "value_key": "wind_direction",
              "trigger_value": 360
          }
      }
  ],

`

DMBlakeley commented on src/tempest.ts:
```diff
@@ -30,6 +33,114 @@ export interface Observation {
+
+  }
+
+  private setTempestData(data): void {
```
Replace lines 90 through 114 with:

`  private setTempestData(data): void {

    const obs = data.obs[0];
    // const windLull = (obs[1] !== null) ? obs[1] : 0;
    const windSpeed = (obs[2] !== null) ? obs[2] * 2.2369 : 0; // convert to mph for heatindex calculation
    // const windGust = (obs[3] !== null) ? Math.round(obs[3] * 2.236936) : 0; // convert to mph
    const T = (obs[7] * 9/5) + 32; // T in F for heatindex, feelsLike and windChill calculations

    // eslint-disable-next-line max-len
    const heatIndex = -42.379 + 2.04901523*T + 10.14333127*obs[8] - 0.22475541*T*obs[8] - 0.00683783*(T**2) - 0.05481717*(obs[8]**2) + 0.00122874*(T**2)*obs[8] + 0.00085282*T*(obs[8]**2) - 0.00000199*(T**2)*(obs[8]**2);

    // feels like temperature on defined for temperatures between 80F and 110F
    const feelsLike = ((T >= 80) && (T <= 110)) ? heatIndex : T;

    // windChill only defined for wind speeds > 3 mph and temperature < 50F
    const windChill = ((windSpeed > 3) && (T < 50)) ? (35.74 + 0.6215*T - 35.75*(windSpeed**0.16) + 0.4275*T*(windSpeed**0.16)) : T;

    this.data = {
      air_temperature: obs[7],
      feels_like: 5/9 * (feelsLike - 32), // convert back to C
      wind_chill: 5/9 * (windChill - 32), // convert back to C
      dew_point: obs[7] - ((100 - obs[8]) / 5.0), // Td = T - ((100 - RH)/5)
      relative_humidity: obs[8],
      wind_avg: obs[2],
      wind_gust: obs[3],
      barometric_pressure: obs[6],
      precip: obs[12],
      precip_accum_local_day: obs[12],
      wind_direction: obs[4],
      solar_radiation: obs[11],
      uv: obs[10],
      brightness: obs[9],
    };
    this.tempest_battery_level = Math.round((obs[16] - 1.8) * 100); // 2.80V = 100%, 1.80V = 0%

  }
`

DMBlakeley committed:
Update package-lock.json

DMBlakeley committed:
v4.0.0 updates

- Fixed error in config.schema.json which was blocking display of occupancy sensor type.
- Added note in README.md to indicate that `precip_accum_local_day` not available in local API.

DMBlakeley commented:
- Fixed additional error in `config.schema.json` which was blocking selection of occupancy sensor types.
- Added note in README.md to indicate that `precip_accum_local_day` is not available when using local_API. Still need to code to check for this condition.

Changes are in my files.

DMBlakeley committed:
Update config.schema.json

Add description that states that  `precip_accum_local_day` is not supported using Local API.

DMBlakeley commented on config.schema.json:
```diff
@@ -1,182 +1,214 @@
+                                  ],
+                                  "condition": {
+                                      "functionBody": "if (!model.local_api && model.sensors[arrayIndices] && model.sensors[arrayIndices].sensor_type && model.sensors[arrayIndices].sensor_type === 'Temperature Sensor') { return true; } else { return false; };"
+                                  }
```
Replace line 185 with:

},
 "description": "Note: `precip_accum_local_day` not supported when using Local API."

chasenicholl commented:
@DMBlakeley thank you for the thoughtful feedback. Please feel free to checkout this branch locally, commit your changes and push here for one more round of reviews before we publish the big update!

edit* apologies I didn't realize I hadn't added you as an official collaborator yet. I just corrected that. You should now be able to directly contribute. Thank you again.

DMBlakeley committed:
Update config.schema.json

DMBlakeley committed:
Check for 'precip_accum_local_day'

Disable `precip_accum_local_day` Occupancy Sensor if using Local API.

DMBlakeley commented:
Hi Chase,
As I am new to being an official collaborator I just wanted to make sure that I am adding or updating files correctly. I see 2 options:

1. Open the `local-websocket-support` branch and add / replace the files directly.
2. Create a new branch as a pull request.

Which do you prefer?
Regards, Doug

chasenicholl commented:
@DMBlakeley great question. Lets have you create a branch from this branch. Make your changes. And open a new PR for me to review merging your changes back into this branch. Make sense?

DMBlakeley commented:
@chasenicholl, I think that creating a PR against the `local-wesocket-support` branch is the safest and cleanest approach. 

Let me give my updates one more review and I will submit the PR for your review and merging into this branch and then back into `main`.

DMBlakeley committed:
Update tempest.ts

DMBlakeley commented:
@chasenicholl, I submitted PR #17 against the Local Socket API branch last evening. Let me know if there are any issues with this approach.

DMBlakeley committed:
Update package-lock.json

chasenicholl committed:
Merge branch 'local-websocket-support' into main

chasenicholl committed:
Merge pull request #17 from DMBlakeley/main

v4.0 review updates

chasenicholl committed:
update readme language to highlight local API support more clearly

chasenicholl commented:
@DMBlakeley I've been running this branch locally now since I merged your PR without issues. How about yourself?

DMBlakeley commented:
@chasenicholl, I have also been running v4.0.0 without issues. Have found very good correlation between v4.0.0 and the Tempest app. My recommendation would be to release.
